components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/choreo/ClientManager.java

```diff
@@ -0,0 +1,175 @@
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+
+import java.io.IOException;
+import java.security.KeyManagementException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.net.ssl.SSLContext;
+
+/**
+ * Class to retrieve the HTTP Clients.
+ */
+public class ClientManager {
+
+    private static final Log LOG = LogFactory.getLog(ClientManager.class);
+
+    private static Map<Integer, CloseableHttpAsyncClient> clientMap = new HashMap<>();
+
+    private PoolingNHttpClientConnectionManager poolingHttpClientConnectionManager;
+
+    private static int HTTP_CONNECTION_TIMEOUT = 300;
+    private static int HTTP_READ_TIMEOUT = 300;
+    private static int HTTP_CONNECTION_REQUEST_TIMEOUT = 300;
+    private static int DEFAULT_MAX_CONNECTIONS = 20;
+
+    public ClientManager() throws FrameworkException {
+
+        poolingHttpClientConnectionManager = createPoolingConnectionManager();
+
+    }
+
+    /**
+     * Get HTTPClient properly configured with tenant configurations.
+     *
+     * @param tenantDomain tenant domain of the service provider.
+     * @return HttpClient
+     */
+    public CloseableHttpAsyncClient getClient(String tenantDomain) throws FrameworkException, IOException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        CloseableHttpAsyncClient client = clientMap.get(tenantId);
+
+        if (client == null) {
+
+            RequestConfig config = createRequestConfig();
+
+            HttpAsyncClientBuilder httpClientBuilder = HttpAsyncClients.custom().setDefaultRequestConfig(config);
+
+            addSslContext(httpClientBuilder, tenantDomain);
+            httpClientBuilder.setConnectionManager(poolingHttpClientConnectionManager);
+
+            client = httpClientBuilder.build();
+            client.start();
+            clientMap.put(tenantId, client);
+        }
+
+        return client;
+    }
+
+    private RequestConfig createRequestConfig() {
+
+        return RequestConfig.custom()
+                .setConnectTimeout(HTTP_CONNECTION_TIMEOUT)
+                .setConnectionRequestTimeout(HTTP_CONNECTION_REQUEST_TIMEOUT)
+                .setSocketTimeout(HTTP_READ_TIMEOUT)
+                .build();
+    }
+
+    private PoolingNHttpClientConnectionManager createPoolingConnectionManager() throws FrameworkException {
+
+        String maxConnectionsString = IdentityUtil.getProperty(Constants.CONNECTION_POOL_MAX_CONNECTIONS);
+        String maxConnectionsPerRouteString = IdentityUtil.getProperty(Constants
+                .CONNECTION_POOL_MAX_CONNECTIONS_PER_ROUTE);
+        int maxConnections = DEFAULT_MAX_CONNECTIONS;
+        int maxConnectionsPerRoute = DEFAULT_MAX_CONNECTIONS;
+        try {
+            maxConnections = Integer.parseInt(maxConnectionsString);
+        } catch (NumberFormatException e) {
+            // Ignore. Default value is used.
```

Comment:
Since we are logging it, are we actually ignoring?

Reply:
fixed